### PR TITLE
Code generator

### PR DESCRIPTION
**Pull request by asymmetric**
Adds a script to run the code generator under `hack/`.

The code generator itself must be installed separately, as explained in `update-codegen.sh`.

### CODE REVIEW

**Comment by zeenix** [comment]
What kind of code this generates?

**Comment by asymmetric** [comment]
See https://blog.openshift.com/kubernetes-deep-dive-code-generation-customresources/

It generates deepcopy functions for each of the types we define, since go has no generics.

**Comment by asymmetric** [comment]
PTALA.

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors.
```
Comment (LiliC) [inline hack/update-codegen.sh]: Not sure about this copyright? Did you mean Chef?

File: Makefile [hunk 2]
```diff
@@ -40,4 +40,7 @@ update-version:
 	find examples helm -regex '.*\.ya?ml$$.bak' -type f \
 		-exec rm '{}' \;
 
+codegen:
+	CODEGEN_PKG=../../../k8s.io/code-generator hack/update-codegen.sh
```
Comment (LiliC) [inline Makefile]: Can't you just do `CODEGEN_PKG=$GOPATH/src/k8s.io/code-generator`?

Comment (asymmetric) [reply Makefile]: No, it has to be a relative path. :man_facepalming:

**Comment by LiliC** [comment]
My only concern is the following:

I tried running this and it seems like I have an older version of the script as I didn't recently update the code-generator repo, so there was a difference when I ran it. I see two options:
1. Commit the script to the repo so we all run the same one.
2. Specify/remind in the README to update to latest version of the code-generator repo.

Up to you which one you choose. :)

**Comment by asymmetric** [comment]
I would exclude 1. because the script gets updated somewhat regularly. I see 2 options:

* Run the `verify-codegen` script with by cloning the latest codegen in Travis, so that the build fails if versions differ
* Add the codegen as a git submodule to the repo (not sure I really want to deal with this one)

**Comment by asymmetric** [comment]
OTOH, if there are no other problems, I think it would make sense to merge this now, so that I can use it in the other outstanding PRs. I would then add a separate ticket to make it more robust.

**Pull request merged**